hooks/pre/35-change-evr-extension-ownership.rb

```diff
@@ -0,0 +1,24 @@
+# In Katello 4.14, the 'evr' extension is removed from PostgreSQL and integrated into the Katello database via a migration.
+# This hook ensures the 'evr' extension's ownership is transferred to the 'foreman' user so migrations can act on it.
+
+if local_postgresql? && execute("rpm -q postgresql-evr", false, false)
+  is_postgresql_active = execute_command("systemctl is-active postgresql", false, true)&.first&.strip == "active"
+
+  # Ensure the PostgreSQL service is running
+  unless is_postgresql_active || app_value(:noop)
+    logger.debug("Starting postgresql service")
+    start_services(['postgresql'])
+  end
+
+  # Update the ownership of the evr extension
+  unless app_value(:noop)
+    logger.debug("Updating ownership of the evr extension if it is enabled")
+    execute!("runuser -l postgres -c \"psql -d foreman -c \\\"UPDATE pg_extension SET extowner = (SELECT oid FROM pg_authid WHERE rolname='foreman') WHERE extname='evr';\\\"\"", false, true)
```

Comment:
First of all, it can be simplified to use `execute_as!`. And I think you should look at the answers as well. Then the final version is:
```suggestion
    database = param_value('foreman', 'db_database')
    username = param_value('foreman', 'db_username')
    sql = "UPDATE pg_extension SET extowner = (SELECT oid FROM pg_authid WHERE rolname='#{username}') WHERE extname='evr'"
    execute_as!('postgres', "psql -d #{database} -c \"#{sql}\"", false, true)
```

But escaping will be tricky, so I'd enhance `execute_as!` to pass it as an array:
```ruby
def execute_as!(user, command, do_say = true, do_log = true)
  runuser_command = ['runuser', '-l', user, '-c', command] 
  execute!(runuser_command, do_say, do_log)
end
```

Then I'd expect it to work if users chose a different DB and/or username as well.

Reply:
I just realized that `Kafo::PuppetCommand.format_command` probably doesn't deal with commands as arrays.

Reply:
Updated to do this:
```
  database = param_value('foreman', 'db_database') || 'foreman'
  username = param_value('foreman', 'db_username') || 'foreman'
  sql = "runuser -l postgres -c \"psql -d '#{database}' -c \\\"UPDATE pg_extension SET extowner = (SELECT oid FROM pg_authid WHERE rolname='#{username}') WHERE extname='evr';\\\"\""
  logger.debug("Executing: #{sql}")
  execute!(sql, false, true) unless app_value(:noop)
```

Still using execute! as escaping was being really painful.

Reply:
I opened https://github.com/theforeman/kafo/pull/374 for that support.

Reply:
Unresolving this and I am going to push an update to use this cause I was running into some escaping issues on the sql command execution.